【提出物】オブジェクト指向版wcコマンド

SeijiNumata opened the pull request:
rubocop,動作検証済み
 Please enter the commit message for your changes. Lines starting

SeijiNumata committed:
wcコマンドを完成

rubocop,動作検証済み
 Please enter the commit message for your changes. Lines starting

SeijiNumata committed:
リファクタリング

kakutani commented on 09.wc_object/main.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require 'optparse'
+require_relative './lib/calc'
```
`'./'`は不要だと思います(他の`require_relative`も同様)

SeijiNumata replied:
手元で試して見たところ、``./``は不要でした。
どこかで間違えて知識を仕入れてしまったようです。
こちら修正しました。ご指摘いただきありがとうございます。

kakutani commented on 09.wc_object/main.rb:
```diff
@@ -0,0 +1,11 @@
+
+options = ARGV.getopts('l')
+wc_command = WcCommand.new
+wc_command.run_wc(options)
```
特にこだわりがなければ、Rubyではこの手の「実行する」というメソッドの名前は `call` で良いと思います。(`wc`という名前が冗長だなと思いました)

SeijiNumata replied:
wcは自明なので確かに冗長ですね。こちらcallにいたしました。

kakutani commented on 09.wc_object/lib/wc_command.rb:
```diff
@@ -0,0 +1,12 @@
+
+class WcCommand
+  def check_argv
+    !ARGV[0].nil?
```
`ARGV`はプロセスの外から渡ってくるものなので、せっかくプロセスの境界として`main.rb`を用意してあるのだから、そちらで処理したほうがよさそうです。

SeijiNumata replied:
こちらmain.rbに処理させるようにいたしました。
外から渡ってきたものは全てmain.rbで処理した方が良いですね。

kakutani commented on 09.wc_object/lib/wc_command.rb:
```diff
@@ -0,0 +1,12 @@
+  end
+
+  def run_wc(option)
+    calculator = check_argv ? FilesCalculator.new(option) : InputCalculator.new(option)
```
`check_argv` という名前だと「プロセスに渡された引数をチェックする」という意味になりますが、何をチェックしているのかを読みとりづらいです。

SeijiNumata replied:
確かにそうですね。
`exists_argv`と命名を変更しましたが、いかがでしょうか。

kakutani commented on 09.wc_object/lib/wc_command.rb:
```diff
@@ -0,0 +1,12 @@
+
+  def run_wc(option)
+    calculator = check_argv ? FilesCalculator.new(option) : InputCalculator.new(option)
+    calculator.run
```
特別な意図がなければ `run`よりも `call`で良さそう。
(`calculator`なんだから「計算するんだろうな」ということは予想できる)

SeijiNumata replied:
承知いたしました。
「一般的にはcallを使う」ということを知らなかったので勉強になりました。

kakutani replied:
好きな名前をつけて構わないのですが、Rubyでは`Proc#call(*arg)`という名前が標準ライブラリで採用されているので、「呼び出す」以上の意図がなければ、`#call`でいいんじゃないかな、というのが言いたいことでした。

必要があれば別の名前を選ぶのは何の問題もありません!

kakutani commented on 09.wc_object/lib/input_calculator.rb:
```diff
@@ -0,0 +1,16 @@
+  end
+
+  def run
+    inputs = $stdin.readlines
```
`main` → `WcCommand` → `InputCalculator` と進んできてここで標準入力から直接読むのは悩ましいですね…。この課題ではインタラクティブなやりとりではなく、標準入力は一度読み込んでしまって文字列として扱えばいいだけなので、`main`で処理してしまうのがいいかもしれません。

`main`はプロセス外部との境界なので、`main`で`ARGV`の状況を判定して、標準入力をまとめて読んでしまったほうが、この課題ではわかりやすいかな、と思います。

SeijiNumata replied:
確かにそうですね。`main.rb` でARGVや標準入力を全てまとめて読むようにいたしました。
その結果、「標準入力があるかどうかを見て、次の処理を決める」という`WcCommand`の存在意義がなくなったため、削除いたしました。

kakutani reviewed (changes requested):
`main.rb` が外部との入出力の境界になっているので、
入力をハンドリングしてRubyオブジェクトに渡して処理をして、
その結果を出力する、という担当にしたほうがよさそうです。

各Calculatorが処理の途中で`print`や`puts`を使って標準出力に都度出力しているようですが、
今回については、
出力させる文字列を組み立てて、その結果を`main.rb`の最後で出力する、
という整理のほうが良さそうに思いました。

参考にしてみてください。

SeijiNumata committed:
レビュー反映　requireの先頭の./が不要だったため削除

SeijiNumata committed:
mainで出力するように設計を変更

SeijiNumata committed:
メソッド名が分かりにくいところを、変更

kakutani commented:
対応していただいようなコメントをいくつかいただいていますが、pushされていないように見えます…。 @SeijiNumata

SeijiNumata committed:
calculatorでハッシュを返し、mainでのformatが簡単になるように修正

mainで出力するように処理を変更しました。

SeijiNumata reviewed (commented):
すみません。
「コメントは送信せずにpendingにしている」と思っていたのですが、送られていたようです。
githubのコメント投稿機能に慣れていなくて、うまく扱えていませんね（困りました😅）

レビューありがとうございます！

修正が完了いたしましたのでpushいたしました。
大きな変更点は以下２点です。
- main.rbに出力機能を持たせる(修正前は各calculatorでputsまでしていたが、修正後は各calculatorは計算結果をハッシュテーブルで返すのみ)
- main.rbで、入力を処理する。(修正前は`wc_command`やcalculatorで処理をしていた)

小さな変更点としては
- メソッドの命名を見直し

などがあります。
ご指摘いただいた内容は全て修正したと思いますが、不足があればご教授いただければ幸いです。

ご確認よろしくお願いします。

kakutani commented:
> 「コメントは送信せずにpendingにしている」と思っていたのですが、送られていたようです。
> githubのコメント投稿機能に慣れていなくて、うまく扱えていませんね（困りました😅）

慣れないとレビューのUIはちょっと難しいですよね。承知しました。

kakutani commented on 09.wc_object/main.rb:
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'optparse'
+require_relative 'lib/calc'
```
`Calc`を直接使っていないので、この行は不要です。

SeijiNumata replied:
全てmain.rbでファイルを呼び出すものだと思っていました。
必要なクラスで必要なファイルを呼び出すように変更いたしました。

kakutani commented on 09.wc_object/lib/input_calculator.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
```
`Calc`を直接使っているので、`require_relative`で読み込んでください。

SeijiNumata replied:
こちら修正いたしました。
ありがとうございます。

kakutani commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
```
`Calc`を直接使っているので、`require_relative`で読み込んでください。

SeijiNumata replied:
こちら修正いたしました。
ありがとうございます。

kakutani reviewed (commented):
`main.rb` ですが、これだけ役割があるのなら、ユーザーからの入力に応じたword countの処理を呼び出すクラス(たとえば `Wc`)にまとめてよさそうです。
`Wc.new.call(options)`として、mailnから呼び出すようなイメージです。

また、この場合の `formatter_lopt` ,`formatter_not_lopt`は、直接'print`するのではなく、表示する出力文字列を組み立てて、`puts Wc.new.call(options)` のような設計にすれば、`Wc`オブジェクトのユニットテストを書く場合などに扱いやすくなります。

SeijiNumata committed:
第二回レビュー反映

wcクラスを定義して、そこでユーザーからの入力に応じたword countの処理を呼び出すように変更
wcクラスで表示する出力文字列を組み立て、mainでputsするように変更
requireを全て一括でmain.rbで呼び出していたため、必要なクラスに書くように修正

SeijiNumata commented:
Wcクラスを定義し、main.rbから呼び出すようにいたしました。
また、` formatter_lopt` ,`formatter_not_lopt`で`print`までしていたため、文字列の組み立てまでに抑えて、main.rbで`puts Wc.new.call(options) `とできるようにいたしました。
確かにこの方がユニットテストが書きやすし、色々と便利そうです。

ご確認お願いいたします。

kakutani commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def call
+    files = validate_and_get_file(ARGV)
```
ARGVはユーザーからの入力なので、ここで直接扱わないのがよさそうです。
`Wc`から必要な値を渡すようにしてください。

SeijiNumata replied:
確かにそうですね。ユーザーからの入力は全てWcで処理しているので、ARGVもWcから渡してあげるようにいたしました。

kakutani commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def calc_depend_opt(str)
+    calc = Calc.new
```
`Calc.new(str)` して、`calc`が自分のインスタンスを元に行数/語数/バイト数を数えるのがよさそうです。
`calc.calc_xxxx`は冗長だと思うので、`calc.xxxx`で良さそうです。

SeijiNumata replied:
なるほど〜〜
この方が毎回メソッドに渡す値を意識しなくて良いのでとても分かりやすいです。

命名についても参考になります。

kakutani commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,64 @@
+    else
+      words = calc.calc_words(str)
+      bytes = calc.calc_bytes(str)
+      row = { lines: lines.to_i, words: words.to_i, bytes: bytes.to_i }
```
calcから返ってくる値が Integerであることを保証してあげると良いのではないでしょうか(都度 `to_i` は冗長にみえます)

SeijiNumata replied:
確かにそうですね。calcメソッド内でintegerが変えるように変更いたしました。

kakutani commented on 09.wc_object/main.rb:
```diff
@@ -0,0 +1,9 @@
+options = ARGV.getopts('l')
+
+wc = Wc.new
+puts wc.call(options)
```
`wc`変数は不要ですね。

```suggestion
puts Wc.new.call(options)
```

SeijiNumata replied:
この書き方知らなかったです。
ありがとうございます。

kakutani commented on 09.wc_object/lib/wc.rb:
```diff
@@ -0,0 +1,50 @@
+
+  private
+
+  def exists_argv
+    if !ARGV[0].nil?
+      true
+    else
+      false
+    end
+  end
```
pコマンドライン引数が設定されていなければ`true`、されていれば `false` としたければ、
`!ARGV[0].nil?` の評価結果をそのまま使えばよいです。

`ARGV`や`$stdin`はユーザーからの入力、`puts`はユーザーへの出力です。ユーザーの入出力を担当する箇所はまとめておきたいですが、`Wc`と`main.rb`の役割分担がなやましいですね…。

SeijiNumata replied:
> pコマンドライン引数が設定されていなければtrue、されていれば false としたければ、
!ARGV[0].nil? の評価結果をそのまま使えばよいです。

否定形がつくと分かりにくいような気がしてわざわざメソッド化していましたが、逆に冗長で不要ですね。

> ARGVや$stdinはユーザーからの入力、putsはユーザーへの出力です。ユーザーの入出力を担当する箇所はまとめておきたいですが、Wcとmain.rbの役割分担がなやましいですね…。

私もそう思います。難しいです...。
考えたこととしては、

「オプション」はFilesCalculator、inputCalculatorのどちらでも使うのでmain.rbで処理する。
`ARGV`や`$stdin`はどちらか一方しか使わないため、分岐のあるWcで処理をする。

という風に分けているとしています。

kakutani commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,64 @@
+    total_words = 0
+    total_bytes = 0
+
+    table.each_with_index do |field, _index|
```
indexを使わないなら、単なる`each`でよさそうです。

SeijiNumata replied:
こちら必要ありませんでした。
単なる`each`に変更いたしました。

SeijiNumata committed:
レビュー第三回反映

受け渡す引数を変えるリファクタリング、
冗長な処理を削除するリファクタリングが主な変更点。
Rubocop、動作確認OK。

komagata commented on 09.wc_object/lib/calc.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class Calc
```
Calculationは略さない方がいいかもです

[プログラミング初心者は変数名やメソッド名を略さない方がいいよ、という話 \- give IT a try](https://blog.jnito.com/entry/2020/10/20/092724)

SeijiNumata replied:
略さないようにしました！

komagata commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require_relative 'calc'
+class FilesCalculator
```
CalcとCalculatorの二つの表現があるのでどちらかに統一した方がいいと思います。

SeijiNumata replied:
Calculatorに統一しました。

komagata commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+  end
+
+  def validate_and_get_file(isfiles)
```
`and`がメソッド名に名前に入るということは二つのことをやっているということなので、1メソッド1つのことをやるようにした方がわかりやすくなると思います〜

SeijiNumata replied:
確かにそうですね。
修正しました！

SeijiNumata committed:
レビュー反映(リファクタリング)、calcの命名とvalidateメソッド

komagata commented on 09.wc_object/lib/calculator.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class Calculator
```
Calculatorだと一般的な名前なので他のプログラムとクラス名がバッティングしちゃうかもです。

moduleを使ったネームスペースの導入を検討してみてください〜

SeijiNumata replied:
なるほど〜、勉強になります。
moduleを利用してみました。
ご確認お願いします。

SeijiNumata committed:
calculatorクラスが一般的な名前のため、モジュールを使ったネームスペースを導入

komagata commented on 09.wc_object/lib/files_calculator.rb:
```diff
@@ -0,0 +1,66 @@
+
+require_relative 'calculator'
+
+class FilesCalculator
```
全てのクラスを同じくネームスペースの下に入れちゃうのがいいと思います〜

SeijiNumata committed:
全てのクラスをネームスペース内に格納
